src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs

```diff
@@ -470,86 +461,62 @@ internal unsafe SocketError DoOperationReceiveFromMultiBuffer(SafeSocketHandle h
         internal unsafe SocketError DoOperationReceiveMessageFrom(Socket socket, SafeSocketHandle handle)
         {
             // WSARecvMsg uses a WSAMsg descriptor.
-            // The WSAMsg buffer is pinned with a GCHandle to avoid complicating the use of Overlapped.
+            // The WSAMsg buffer is a pinned array to avoid complicating the use of Overlapped.
             // WSAMsg contains a pointer to a sockaddr.
-            // The sockaddr is pinned with a GCHandle to avoid complicating the use of Overlapped.
+            // The sockaddr is a pinned array to avoid complicating the use of Overlapped.
             // WSAMsg contains a pointer to a WSABuffer array describing data buffers.
             // WSAMsg also contains a single WSABuffer describing a control buffer.
-            PinSocketAddressBuffer();
 
             // Create a WSAMessageBuffer if none exists yet.
-            if (_wsaMessageBuffer == null)
+            if (_wsaMessageBufferPinned == null)
             {
-                Debug.Assert(!_wsaMessageBufferGCHandle.IsAllocated);
-                _wsaMessageBuffer = new byte[sizeof(Interop.Winsock.WSAMsg)];
-            }
-
-            // And ensure the WSAMessageBuffer is appropriately pinned.
-            Debug.Assert(!_wsaMessageBufferGCHandle.IsAllocated || _wsaMessageBufferGCHandle.Target == _wsaMessageBuffer);
-            if (!_wsaMessageBufferGCHandle.IsAllocated)
-            {
-                _wsaMessageBufferGCHandle = GCHandle.Alloc(_wsaMessageBuffer, GCHandleType.Pinned);
+                _wsaMessageBufferPinned = GC.AllocateUninitializedArray<byte>(sizeof(Interop.Winsock.WSAMsg), pinned: true);
```

Comment:
Nit: this can be simplified now to:
```C#
_wsaMessageBufferPinned ??= GC.AllocateUninitializedArray<byte>(sizeof(Interop.Winsock.WSAMsg), pinned: true);
```